src/firmware_new/boards/frankie_v1/io/ublox_odinw262_communicator.c

```diff
@@ -0,0 +1,209 @@
+#include "firmware_new/boards/frankie_v1/io/ublox_odinw262_communicator.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "FreeRTOS.h"
+#include "cmsis_os.h"
+#include "firmware/app/logger/logger.h"
+#include "firmware_new/boards/frankie_v1/io/gpio_pin.h"
+#include "firmware_new/boards/frankie_v1/io/ublox_odinw262_communicator_buffer_utils.h"
+#include "firmware_new/boards/frankie_v1/usart.h"
+#include "main.h"
+#include "task.h"
+
+// TODO have a way of configuring these values externally
+// https://github.com/UBC-Thunderbots/Software/issues/1876
+#define WIFI_SSID "thunderbots"
+#define WIFI_PASS "tbots1234"
+
+#define DMA_BUFFER __attribute__((section(".dma_buffer")))
+#define RX_BUFFER_LENGTH_BYTES 4096
+
+// Buffers: The DMA_BUFFER is managed by the UART peripheral configured in "circular" mode
+DMA_BUFFER static uint8_t g_dma_uart_receive_buffer[RX_BUFFER_LENGTH_BYTES] = {0};
+static char g_uart_receive_buffer[RX_BUFFER_LENGTH_BYTES]                   = {0};
+
+// Interrupt state: the response status and counters are used by handleidleline
+volatile size_t g_dma_counter_on_uart_idle_line        = 0;
+volatile size_t g_last_byte_parsed_from_dma_buffer     = 0;
+volatile UbloxResponseStatus_t g_ublox_response_status = UBLOX_RESPONSE_UNDETERMINED;
+
+static osSemaphoreId_t g_dma_receive_semaphore;
+static UART_HandleTypeDef* g_ublox_uart_handle;
+static GpioPin_t* g_ublox_reset_pin;
+static uint32_t g_ublox_response_timeout;
+
+static bool g_initialized = false;
```

Comment:
Can we reduce the number of global variables? e.g. Can the user of these functions keep track of a struct?

Reply:
These are local to the `ublox_odinw262_communicator.c` file. Putting them in a struct doesn't really make a difference because dependency injection is a pain for the IO layer. The `usart.c` file is auto-generated and we need to call the `io_ublox_communicator_handleIdleLine` function there. Its really hard to pass that struct to that file and into the interrupt. 

So in that case, the struct itself would have to be static, which is not that far off from static variables defined at the top.